File: dynamic-programming/unique-paths/unique-paths.go
package uniquepaths

func uniquePaths(m int, n int) int {
	dp := make([][]int, n)

	for i := 0; i < m; i++ {
		dp[0] = append(dp[0], 1)
	}

	for i := 1; i < n; i++ {
		dp[i] = append(dp[i], 1)
	}

	for i := 1; i < n; i++ {
		for j := 1; j < m; j++ {
			top := dp[i-1][j]
			left := dp[i][j-1]

			dp[i] = append(dp[i], top+left)
		}
	}

	return dp[n-1][m-1]
}

File: dynamic-programming/unique-paths/unique-paths.ts
function uniquePaths(m: number, n: number): number {
  const dp: number[][] = new Array(m).fill(null).map(() => new Array(n));

  for (let i = 0; i < n; i++) {
    dp[0][i] = 1;
  }

  for (let i = 0; i < m; i++) {
    dp[i][0] = 1;
  }

  for (let i = 1; i < m; i++) {
    for (let j = 1; j < n; j++) {
      let top: number = dp[i - 1][j];
      let left: number = dp[i][j - 1];
      dp[i][j] = top + left;
    }
  }

  return dp[m - 1][n - 1];
}
